### ISSUE TITLE
[Phase 7] Integration & CLI

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/dynamic-clv-segmentation#8]
## Summary

Create the end-to-end pipeline orchestration and command-line interface. This phase integrates all components into a usable system with sample data and documentation.

**Priority:** 🟢 Final polish - Integration and UX
**Parent Issue:** #1
**Depends On:** All previous phases

---

## Objectives

1. Build unified pipeline class combining all components
2. Create CLI for training, prediction, and model management
3. Provide sample dataset and demo script
4. Write end-to-end integration tests

---

## Files to Create

| File | Lines (est.) | Description |
|------|--------------|-------------|
| `src/pipeline.py` | 250 | Full pipeline orchestration |
| `src/cli.py` | 200 | Command-line interface |
| `tests/test_integration.py` | 200 | E2E tests |
| `examples/demo.py` | 80 | Demo script |
| `examples/sample_data.csv` | N/A | Sample transactions |
| `pyproject.toml` | 60 | Updated with CLI entry point |

---

### Pipeline Class
- [ ] Combine data loading, features, segmentation, prediction
- [ ] Single `fit()` method trains full pipeline
- [ ] `predict()` returns CLV with confidence intervals
- [ ] `save()`/`load()` for model persistence
- [ ] Configuration via dataclass

### CLI
- [ ] `train` command: fit pipeline on data
- [ ] `predict` command: score new customers
- [ ] `evaluate` command: compute metrics on test set
- [ ] `list-models` command: show saved versions
- [ ] Helpful error messages and --help

### Sample Data
- [ ] Generate synthetic transaction data
- [ ] Include various customer patterns
- [ ] Document expected outputs

### Integration Tests
- [ ] Full train -> predict cycle
- [ ] Model save -> load -> predict
- [ ] CLI commands work

---

## Starter Code

📄 **File:** `src/pipeline.py`

```python
"""End-to-end CLV prediction pipeline."""

from dataclasses import dataclass, field
from pathlib import Path
import numpy as np
import pandas as pd

from src.features.rfm import RFMExtractor
from src.features.sequence_encoder import TransactionSequenceEncoder
from src.segmentation.gmm import CustomerSegmenter
from src.prediction.stacker import StackedCLVPredictor
from src.uncertainty.bootstrap import BootstrapPredictor
from src.persistence.model_store import ModelStore, compute_data_hash


@dataclass
class PipelineConfig:
    """Configuration for CLV prediction pipeline."""
    # Data columns
    customer_id_col: str = "customer_id"
    transaction_date_col: str = "transaction_date"
    monetary_col: str = "amount"

    # Segmentation
    segment_k_range: tuple[int, int] = (2, 8)

    # Prediction
    stacker_n_folds: int = 5

    # Uncertainty
    n_bootstrap: int = 100
    confidence_level: float = 0.95

    # Persistence
    model_dir: str = "models"


@dataclass
class CLVPrediction:
    """Container for CLV prediction results."""
    customer_id: str
    segment_id: int
    segment_probabilities: np.ndarray
    clv_point_estimate: float
    clv_lower_bound: float
    clv_upper_bound: float


class CLVPipeline:
    """
    Complete pipeline for Customer Lifetime Value prediction.

    Components:
    1. RFM feature extraction
    2. Transaction sequence embeddings
    3. GMM-based customer segmentation
    4. Stacked ensemble CLV prediction
    5. Bootstrap confidence intervals
    """

    def __init__(self, config: PipelineConfig | None = None):
        self.config = config or PipelineConfig()

        # Initialize components
        self._rfm_extractor = RFMExtractor()
        self._sequence_encoder = TransactionSequenceEncoder(
            customer_id_col=self.config.customer_id_col,
            transaction_date_col=self.config.transaction_date_col,
            monetary_col=self.config.monetary_col
        )
        self._segmenter = CustomerSegmenter(k_range=self.config.segment_k_range)

        # Will be initialized during fit
        self._predictor: BootstrapPredictor | None = None
        self._model_store = ModelStore(self.config.model_dir)
        self._is_fitted = False

    def fit(
        self,
        transactions: pd.DataFrame,
        target_clv: pd.Series
    ) -> "CLVPipeline":
        """
        Fit the complete pipeline.

        Args:
            transactions: Transaction-level data
            target_clv: Actual CLV per customer (for training)
        """
        # Extract features
        rfm_features = self._rfm_extractor.fit_transform(transactions)
        embeddings = self._sequence_encoder.fit_transform(transactions)

        # Combine features
        customer_ids = rfm_features.index
        X = np.hstack([
            rfm_features.values,
            embeddings
        ])

        # Fit segmenter
        self._segmenter.fit(X)
        segment_probs = self._segmenter.predict_proba(X)

        # Align target with features
        y = target_clv.loc[customer_ids].values

        # Fit predictor with bootstrap
        def make_predictor():
            return StackedCLVPredictor(
                n_folds=self.config.stacker_n_folds
            )

        self._predictor = BootstrapPredictor(
            base_predictor_factory=make_predictor,
            n_bootstrap=self.config.n_bootstrap,
            confidence_level=self.config.confidence_level
        )
        self._predictor.fit(X, y, segment_probs=segment_probs)

        self._is_fitted = True
        return self

    def predict(self, transactions: pd.DataFrame) -> list[CLVPrediction]:
        """Predict CLV for customers in transaction data."""
        if not self._is_fitted:
            raise ValueError("Must call fit() first")

        # Extract features
        rfm_features = self._rfm_extractor.transform(transactions)
        embeddings = self._sequence_encoder.transform(transactions)

        customer_ids = rfm_features.index.tolist()
        X = np.hstack([rfm_features.values, embeddings])

        # Get segments
        segment_probs = self._segmenter.predict_proba(X)
        segment_labels = self._segmenter.predict(X)

        # Get predictions with intervals
        intervals = self._predictor.predict_with_intervals(
            X, segment_probs=segment_probs
        )

        # Package results
        results = []
        for i, cid in enumerate(customer_ids):
            results.append(CLVPrediction(
                customer_id=cid,
                segment_id=int(segment_labels[i]),
                segment_probabilities=segment_probs[i],
                clv_point_estimate=intervals[i].point_estimate,
                clv_lower_bound=intervals[i].lower_bound,
                clv_upper_bound=intervals[i].upper_bound
            ))

        return results

    def save(self, metrics: dict | None = None) -> str:
        """Save fitted pipeline, return version ID."""
        # Placeholder: would save all components
        raise NotImplementedError

    def load(self, version_id: str | None = None) -> "CLVPipeline":
        """Load pipeline from saved version."""
        raise NotImplementedError
```

📄 **File:** `src/cli.py`

```python
"""Command-line interface for CLV prediction."""

import argparse
import sys
from pathlib import Path

import pandas as pd

from src.pipeline import CLVPipeline, PipelineConfig


def cmd_train(args):
    """Train the CLV pipeline."""
    print(f"Loading data from {args.data}...")
    transactions = pd.read_csv(args.data)

    if args.target:
        target = pd.read_csv(args.target, index_col=0).squeeze()
    else:
        print("Error: --target required for training")
        sys.exit(1)

    config = PipelineConfig(model_dir=args.output)
    pipeline = CLVPipeline(config)

    print("Fitting pipeline...")
    pipeline.fit(transactions, target)

    print(f"Model saved to {args.output}")


def cmd_predict(args):
    """Generate predictions for new customers."""
    print(f"Loading model from {args.model}...")
    config = PipelineConfig(model_dir=args.model)
    pipeline = CLVPipeline(config)
    pipeline.load()

    print(f"Loading customer data from {args.customers}...")
    transactions = pd.read_csv(args.customers)

    predictions = pipeline.predict(transactions)

    # Output results
    results = pd.DataFrame([
        {
            "customer_id": p.customer_id,
            "segment": p.segment_id,
            "clv_estimate": p.clv_point_estimate,
            "clv_lower": p.clv_lower_bound,
            "clv_upper": p.clv_upper_bound
        }
        for p in predictions
    ])

    if args.output:
        results.to_csv(args.output, index=False)
        print(f"Predictions saved to {args.output}")
    else:
        print(results.to_string())


def cmd_list_models(args):
    """List saved model versions."""
    from src.persistence.model_store import ModelStore
    store = ModelStore(args.model_dir)

    versions = store.list_versions()
    if not versions:
        print("No saved models found.")
        return

    print(f"{'Version':<25} {'Created':<25} {'Samples':<10} {'R2':<10}")
    print("-" * 70)
    for v in versions:
        r2 = v.metrics.get("r2", "N/A")
        print(f"{v.version_id:<25} {v.created_at:<25} {v.n_samples:<10} {r2:<10}")


def main():
    parser = argparse.ArgumentParser(
        description="Dynamic CLV Segmentation CLI"
    )
    subparsers = parser.add_subparsers(dest="command", required=True)

    # Train command
    train_parser = subparsers.add_parser("train", help="Train CLV model")
    train_parser.add_argument("--data", required=True, help="Transaction CSV")
    train_parser.add_argument("--target", required=True, help="Target CLV CSV")
    train_parser.add_argument("--output", default="models", help="Output directory")
    train_parser.set_defaults(func=cmd_train)

    # Predict command
    predict_parser = subparsers.add_parser("predict", help="Generate predictions")
    predict_parser.add_argument("--model", required=True, help="Model directory")
    predict_parser.add_argument("--customers", required=True, help="Customer CSV")
    predict_parser.add_argument("--output", help="Output CSV (optional)")
    predict_parser.set_defaults(func=cmd_predict)

    # List models command
    list_parser = subparsers.add_parser("list-models", help="List saved models")
    list_parser.add_argument("--model-dir", default="models", help="Model directory")
    list_parser.set_defaults(func=cmd_list_models)

    args = parser.parse_args()
    args.func(args)


if __name__ == "__main__":
    main()
```

📄 **File:** `examples/demo.py`

```python
"""Demo script for CLV prediction pipeline."""

import numpy as np
import pandas as pd
from datetime import datetime, timedelta

# Generate synthetic transaction data
def generate_sample_data(n_customers: int = 100, seed: int = 42) -> pd.DataFrame:
    """Generate synthetic transaction data."""
    np.random.seed(seed)

    records = []
    base_date = datetime(2023, 1, 1)

    for cid in range(n_customers):
        customer_id = f"C{cid:04d}"

        # Random customer behavior
        avg_frequency = np.random.exponential(30)  # days between purchases
        avg_amount = np.random.lognormal(4, 0.5)  # ~$55 average
        n_transactions = max(1, int(np.random.poisson(8)))

        current_date = base_date + timedelta(days=np.random.randint(0, 90))

        for _ in range(n_transactions):
            records.append({
                "customer_id": customer_id,
                "transaction_date": current_date.strftime("%Y-%m-%d"),
                "amount": max(1, np.random.lognormal(np.log(avg_amount), 0.3))
            })
            current_date += timedelta(days=max(1, int(np.random.exponential(avg_frequency))))

            if current_date > datetime(2024, 6, 1):
                break

    return pd.DataFrame(records)


if __name__ == "__main__":
    print("Generating sample transaction data...")
    transactions = generate_sample_data(n_customers=200)
    transactions.to_csv("examples/sample_data.csv", index=False)
    print(f"Created {len(transactions)} transactions for {transactions['customer_id'].nunique()} customers")
    print(f"Saved to examples/sample_data.csv")
    print("\nSample:")
    print(transactions.head(10))
```

---

## CLI Usage Examples

```bash
# Train a model
python -m src.cli train \
    --data examples/sample_data.csv \
    --target examples/target_clv.csv \
    --output models/

# Generate predictions
python -m src.cli predict \
    --model models/ \
    --customers new_customers.csv \
    --output predictions.csv

# List saved models
python -m src.cli list-models --model-dir models/
```

---

## Definition of Done

- [ ] `python -m src.cli train` works without errors
- [ ] `python -m src.cli predict` outputs valid CSV
- [ ] Demo script runs and produces sample data
- [ ] Integration tests pass
- [ ] All phase components work together
- [ ] README updated with usage instructions
